core/trino-main/src/main/java/io/trino/sql/gen/columnar/BloomFilter.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.gen.columnar;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import io.airlift.slice.Slice;
+import io.airlift.slice.XxHash64;
+import io.trino.operator.project.InputChannels;
+import io.trino.spi.Page;
+import io.trino.spi.block.VariableWidthBlock;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.predicate.Domain;
+import io.trino.spi.type.CharType;
+import io.trino.spi.type.Type;
+import io.trino.spi.type.VarbinaryType;
+import io.trino.spi.type.VarcharType;
+
+import java.util.List;
+import java.util.function.Supplier;
+
+public class BloomFilter
+{
+    private BloomFilter() {}
+
+    public static boolean canUseBloomFilter(Domain domain)
+    {
+        Type type = domain.getType();
+        if (type.getJavaType() != Slice.class) {
+            return false;
+        }
+        if (!(type instanceof VarcharType || type instanceof CharType || type instanceof VarbinaryType)) {
+            return false;
+        }
+        return !domain.isNone() && !domain.isAll() && domain.isNullableDiscreteSet();
+    }
+
+    public static Supplier<FilterEvaluator> createBloomFilterEvaluator(Domain discreteDomain, int inputChannel)
+    {
+        return () -> new ColumnarFilterEvaluator(
+                new DictionaryAwareColumnarFilter(
+                        new BloomColumnarFilter(discreteDomain.getNullableDiscreteSet(), inputChannel)));
+    }
+
+    private static final class BloomColumnarFilter
+            implements ColumnarFilter
+    {
+        private final SliceBloomFilter bloomFilter;
+        private final boolean isNullAllowed;
+        private final InputChannels inputChannels;
+
+        public BloomColumnarFilter(Domain.DiscreteSet discreteSet, int inputChannel)
+        {
+            this.isNullAllowed = discreteSet.containsNull();
+            this.bloomFilter = new SliceBloomFilter((List<Slice>) (List<?>) discreteSet.getNonNullValues());
```

Comment:
why two casts?

Reply:
unfortunately it doesn't compile any other way

Reply:
We could make DescreteSet a generic type instead but this would be a follow up